include/pf-applications/time_integration/time_marching.h

```diff
@@ -0,0 +1,516 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2025 by the hpsint authors
+//
+// This file is part of the hpsint library.
+//
+// The hpsint library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 3.0 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.MD at
+// the top level directory of hpsint.
+//
+// ---------------------------------------------------------------------
+
+#pragma once
+
+#include <pf-applications/lac/evaluation.h>
+#include <pf-applications/lac/solvers_linear.h>
+#include <pf-applications/lac/solvers_linear_parameters.h>
+#include <pf-applications/lac/solvers_nonlinear.h>
+#include <pf-applications/lac/solvers_nonlinear_parameters.h>
+
+namespace TimeIntegration
+{
+  template <typename VectorType>
+  class TimeMarching
+  {
+  public:
+    virtual ~TimeMarching() = default;
+
+    virtual void
+    make_step(VectorType &) = 0;
+
+    virtual void
+    clear() = 0;
+  };
+
+  using namespace dealii;
+  using namespace hpsint;
+  using namespace NonLinearSolvers;
+
+  template <int dim,
+            typename VectorType,
+            typename NonLinearOperator,
+            typename StreamType>
+  class TimeMarchingImplicit : public TimeMarching<VectorType>
+  {
+  public:
+    using Number = typename NonLinearOperator::value_type;
+
+    TimeMarchingImplicit(
+      NonLinearOperator                           &nonlinear_operator,
+      Preconditioners::PreconditionerBase<Number> &preconditioner,
+      const ResidualWrapper<Number>               &residual_wrapper,
+      const DoFHandler<dim>                       &dof_handler,
+      const AffineConstraints<Number>             &constraints,
+      const NonLinearData                         &nonlinear_params,
+      NewtonSolverSolverControl                   &statistics,
+      MyTimerOutput                               &timer,
+      StreamType                                  &stream,
+      std::function<void(const VectorType &)>      setup_custom_preconditioner,
+      std::function<void(const VectorType &)>      setup_linearization_point,
+      std::function<
+        std::vector<std::tuple<std::string, unsigned int, unsigned int>>(void)>
+        quantities_to_check)
```

Comment:
```suggestion
      const std::function<void(const VectorType &)>      setup_custom_preconditioner,
      const std::function<void(const VectorType &)>      setup_linearization_point,
      const std::function<
        std::vector<std::tuple<std::string, unsigned int, unsigned int>>(void)>
        quantities_to_check)
```

Reply:
The same reasoning as before: I move these lambdas later in the code into the other callbacks.